data/pigui/modules/new-game-window/recovery.lua

```diff
@@ -30,23 +30,37 @@ local fixupDoc = Helpers.versioned {{
 
 		local refs = {}
 
+		local function unpickle(tbl)
+			local result = {}
+			tbl = tbl.table
+			for i = 1, #tbl, 2 do
+				result[tbl[i]] = tbl[i + 1]
```

Comment:
This function could handle references as well:

```lua
local k, v = tbl[i], tbl[i + 1]
if type(k) == "table" and k.ref then k = cache[k.ref] end
if type(v) == "table" and v.ref then v = cache[v.ref] end
result[k] = v
```

It could even be handled another way to combine the `getRefs`, `putRefs`, and `unpickle` functions together:

```lua
local cache = {}

local function unpickle(v)
	if type(v) ~= 'table' or not v.ref then return v end
	if not v.table then return cache[v.ref] end

	local result = {}
	cache[v.ref] = result

	for i = 1, #v.table, 2 do
		local key = unpickle(v.table[i])
		local val = unpickle(v.table[i + 1])
		result[key] = val
	end

	return result
end
```

Reply:
The proposed version of the function will definitely not work, because the function will end on the first line - `gameDoc` does not have a `ref` field.
It is also assumed that the definition of a lua object will necessarily be encountered before other references to it, but this does not seem to be the case, at least because we are dealing with general Lua table at top level.

Reply:
It was meant to illustrate the concept more than be tested/proven-working code. That being said, I see your point about iteration order; although pickled Lua tables internally require that the first encountered instance of a table value contains its definition when unserializing, that may not hold true when operating on recovered save data that's not going through the C++ serialization process.

Reply:
I picked up some things from your code, but I still have to do two passes.